Draft of branch coverage support

Zalathar opened the pull request:
EDIT: I've closed this and re-opened a non-draft version at #122322.

This is a work-in-progress snapshot of my implementation of branch coverage.

~~It works, and produces correct results in many cases. However, it sometimes produces confusing or incorrect results for `if` expressions that contain the `!`/`&&`/`||` operators, since the THIR-to-MIR lowering of those is non-trivial.~~

Zalathar committed:
coverage: Include recorded branch info in coverage instrumentation/codegen

Zalathar commented:
@rustbot author

Zalathar commented on tests/coverage/branch.coverage:
```diff
@@ -0,0 +1,102 @@
+   LL|     15|fn if_or(a: bool, b: bool) {
+   LL|     15|    do_not_merge_with_signature!();
+   LL|       |
+   LL|     15|    if a || b {
+                          ^3
+  ------------------
+  |  Branch (LL:8): [True: 12, False: 3]
+  |  Branch (LL:13): [True: 14, False: 1]
+  ------------------
```
Notice that the *true* count for `b` is wrong here, since it should be 2 instead of 14.

Zalathar commented on tests/coverage/branch.coverage:
```diff
@@ -0,0 +1,120 @@
+   LL|     15|fn if_not(cond: bool) {
+   LL|     15|    do_not_merge_with_signature!();
+   LL|       |
+   LL|       |    if
+   LL|       |        !
+   LL|     15|        cond
+  ------------------
+  |  Branch (LL:9): [True: 10, False: 5]
+  ------------------
```
Two problems here:
- The span doesn't include `!`.
- The true/false counts are reversed, since they're tracking `cond` instead of `!cond`.

Zalathar commented on tests/coverage/branch.coverage:
```diff
@@ -0,0 +1,120 @@
+   LL|     15|fn if_and(a: bool, b: bool) {
+   LL|     15|    do_not_merge_with_signature!();
+   LL|       |
+   LL|       |    if
+   LL|     15|        a
+  ------------------
+  |  Branch (LL:9): [True: 12, False: 3]
+  ------------------
+   LL|       |        &&
+   LL|     12|        b
+  ------------------
+  |  Branch (LL:9): [True: 8, False: 4]
+  ------------------
```
This all looks correct, which is nice.

Zalathar commented on tests/coverage/branch.coverage:
```diff
@@ -0,0 +1,120 @@
+   LL|     15|fn if_or(a: bool, b: bool) {
+   LL|     15|    do_not_merge_with_signature!();
+   LL|       |
+   LL|       |    if
+   LL|     15|        a
+  ------------------
+  |  Branch (LL:9): [True: 12, False: 3]
+  ------------------
+   LL|       |        ||
+   LL|      3|        b
+  ------------------
+  |  Branch (LL:9): [True: 14, False: 1]
+  ------------------
```
Here the line count for `b` is correct, but the branch count's *true* arm is blatantly wrong, since it should be 2 instead of 14.

Zalathar replied:
Ah, I think what's happened here is that the count is (12 + 2 = 14).

The marker for `b == true` gets injected into the *then* block that both `a == true` and `b == true` branch to, so it has the execution count for both paths, which is not what we want.

I need to have the RHS inject its *true* marker into a separate block, which then jumps to the *then* block.

Zalathar commented on tests/coverage/branch.coverage:
```diff
@@ -0,0 +1,120 @@
+   LL|     15|fn if_let(cond: bool) {
+   LL|     15|    do_not_merge_with_signature!();
+   LL|       |
+   LL|       |    if
+   LL|       |        let
+   LL|       |        true
+   LL|       |        =
+   LL|     15|        cond
```
This currently doesn't get branch coverage at all. I'd like to fix that if possible, but for me it's a lower priority than getting the logical ops correct.

Zalathar commented:
cc @Swatinem in case you're curious.

The MIR building stuff will have to change once I figure out how to handle the logical ops better.

bors commented:
:umbrella: The latest upstream changes (presumably #118300) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Zalathar commented on tests/coverage/bad_counter_ids.coverage:
```diff
@@ -40,6 +40,10 @@
    LL|      0|}
    LL|       |
    LL|      1|fn eq_bad_message() {
+  ------------------
+  |  Branch (LL:21): [True: 0, False: 1]
+  |  Branch (LL:21): [True: 1, False: 0]
+  ------------------
```
Ah, I think what's happened here is that these spans are from a completely different file (via macro-expansion), which my branch-span code makes no attempt to deal with. So we just end up with completely bogus coordinates in the mappings.

Zalathar commented on tests/coverage/try_error_result.coverage:
```diff
@@ -39,7 +45,17 @@
    LL|       |struct Thing1;
    LL|       |impl Thing1 {
    LL|     18|    fn get_thing_2(&self, return_error: bool) -> Result<Thing2, ()> {
+  ------------------
+  |  Branch (LL:21): [True: 0, False: 0]
+  |  Branch (LL:21): [True: 0, False: 0]
+  |  Branch (LL:21): [True: 5, False: 0]
+  |  Branch (LL:21): [True: 5, False: 0]
+  |  Branch (LL:21): [True: 5, False: 0]
+  ------------------
```
I think these are also from macro-expansion. The definition of `assert_eq!` must contain an `if` statement on line 40 of its enclosing file, which is why these all end up here.

Zalathar commented:
Well, this definitely indicates a need to add macro-expanded spans to the list of things I need to figure out.

The existing coverage code has some functions in `spans.rs` for dealing with this sort of thing, which I can probably look to for inspiration.

Zalathar commented:
I tried passing each of the branch spans through `function_source_span`. That gets rid of the bogus spans, but it also results in a lot of unhelpful noise for assertions.

Zalathar commented on tests/coverage/branch.coverage:
```diff
@@ -0,0 +1,123 @@
+  |  Branch (LL:5): [True: 15, False: 0]
+  ------------------
+   LL|       |
+   LL|       |    if
+   LL|     15|        a
+  ------------------
+  |  Branch (LL:9): [True: 12, False: 3]
+  ------------------
+   LL|       |        ||
+   LL|      3|        b
+  ------------------
+  |  Branch (LL:9): [True: 2, False: 1]
+  ------------------
+   LL|     14|    {
```
I managed to fix this by injecting a `goto` to a fresh block after injecting the branch markers. This means that when `then_else_break` ties together the two success arms of `||`, neither one gets double-counted.

Zalathar commented:
Now that the `||` bug is fixed, there are no known correctness issues remaining. The `!` mappings are non-ideal, but they're not *wrong*.

So I'm thinking it might make sense for me to polish the current functionality into a review-ready state, so that people can start playing with branch coverage on nightly and start flushing out any lingering correctness issues.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7016451077/job/19087702801) [(plain)](https://github.com/rust-lang/rust/commit/9000afd4b6f8624e2b8b5562ac932a113287738c/checks/19087702801/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=Zalathar
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_05f509a9-910c-4c04-92d6-2fa426b4ddf7
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=5af52c8de2399e9e237df08270d0a8c023547e5d
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_05f509a9-910c-4c04-92d6-2fa426b4ddf7
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_05f509a9-910c-4c04-92d6-2fa426b4ddf7
GITHUB_TRIGGERING_ACTOR=Zalathar
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/118305/merge
GITHUB_WORKFLOW_SHA=5af52c8de2399e9e237df08270d0a8c023547e5d
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
---- [coverage-map] tests/coverage/conditions.rs stdout ----
diff of cov-map:

1 Function name: conditions::main
- Raw bytes (1297): 0x[01, 01, e9, 01, 09, 5f, 63, 41, 67, 3d, 35, 39, 05, 00, a3, 07, 09, 05, 00, a3, 07, 09, 05, 00, 9e, 07, 0d, a3, 07, 09, 05, 00, 0d, 35, 0d, 35, 35, 39, 4a, 39, 0d, 35, 4a, 39, 0d, 35, 67, 3d, 35, 39, 67, 3d, 35, 39, 63, 41, 67, 3d, 35, 39, 9e, 07, 0d, a3, 07, 09, 05, 00, 45, 00, eb, 01, 49, 45, 00, eb, 01, 49, 45, 00, e6, 01, 31, eb, 01, 49, 45, 00, e6, 01, 31, eb, 01, 49, 45, 00, e2, 01, 4d, e6, 01, 31, eb, 01, 49, 45, 00, e2, 01, 4d, e6, 01, 31, eb, 01, 49, 45, 00, 4d, 51, de, 01, 51, e2, 01, 4d, e6, 01, 31, eb, 01, 49, 45, 00, de, 01, 51, e2, 01, 4d, e6, 01, 31, eb, 01, 49, 45, 00, 97, 02, 55, 4d, 51, 97, 02, 55, 4d, 51, 93, 02, 59, 97, 02, 55, 4d, 51, 49, 8f, 02, 93, 02, 59, 97, 02, 55, 4d, 51, 61, 00, 8f, 03, 65, 61, 00, 8f, 03, 65, 61, 00, 8a, 03, 2d, 8f, 03, 65, 61, 00, 8a, 03, 2d, 8f, 03, 65, 61, 00, 86, 03, 69, 8a, 03, 2d, 8f, 03, 65, 61, 00, 86, 03, 69, 8a, 03, 2d, 8f, 03, 65, 61, 00, 69, 6d, 82, 03, 6d, 86, 03, 69, 8a, 03, 2d, 8f, 03, 65, 61, 00, 82, 03, 6d, 86, 03, 69, 8a, 03, 2d, 8f, 03, 65, 61, 00, bf, 03, 71, 69, 6d, bf, 03, 71, 69, 6d, bb, 03, 75, bf, 03, 71, 69, 6d, b3, 03, 00, 65, b7, 03, bb, 03, 75, bf, 03, 71, 69, 6d, 7d, cf, 06, d3, 06, 8d, 01, d7, 06, 89, 01, 81, 01, 85, 01, 79, 00, c7, 04, 7d, 79, 00, c7, 04, 7d, 79, 00, c2, 04, 29, c7, 04, 7d, 79, 00, c2, 04, 29, c7, 04, 7d, 79, 00, be, 04, 81, 01, c2, 04, 29, c7, 04, 7d, 79, 00, be, 04, 81, 01, c2, 04, 29, c7, 04, 7d, 79, 00, 81, 01, 85, 01, ba, 04, 85, 01, be, 04, 81, 01, c2, 04, 29, c7, 04, 7d, 79, 00, ba, 04, 85, 01, be, 04, 81, 01, c2, 04, 29, c7, 04, 7d, 79, 00, d7, 06, 89, 01, 81, 01, 85, 01, d7, 06, 89, 01, 81, 01, 85, 01, d3, 06, 8d, 01, d7, 06, 89, 01, 81, 01, 85, 01, 11, ff, 06, 83, 07, 21, 87, 07, 1d, 15, 19, 7d, cf, 06, d3, 06, 8d, 01, d7, 06, 89, 01, 81, 01, 85, 01, cb, 06, 11, 7d, cf, 06, d3, 06, 8d, 01, d7, 06, 89, 01, 81, 01, 85, 01, cb, 06, 11, 7d, cf, 06, d3, 06, 8d, 01, d7, 06, 89, 01, 81, 01, 85, 01, c6, 06, 25, cb, 06, 11, 7d, cf, 06, d3, 06, 8d, 01, d7, 06, 89, 01, 81, 01, 85, 01, c6, 06, 25, cb, 06, 11, 7d, cf, 06, d3, 06, 8d, 01, d7, 06, 89, 01, 81, 01, 85, 01, c2, 06, 15, c6, 06, 25, cb, 06, 11, 7d, cf, 06, d3, 06, 8d, 01, d7, 06, 89, 01, 81, 01, 85, 01, c2, 06, 15, c6, 06, 25, cb, 06, 11, 7d, cf, 06, d3, 06, 8d, 01, d7, 06, 89, 01, 81, 01, 85, 01, 15, 19, be, 06, 19, c2, 06, 15, c6, 06, 25, cb, 06, 11, 7d, cf, 06, d3, 06, 8d, 01, d7, 06, 89, 01, 81, 01, 85, 01, be, 06, 19, c2, 06, 15, c6, 06, 25, cb, 06, 11, 7d, cf, 06, d3, 06, 8d, 01, d7, 06, 89, 01, 81, 01, 85, 01, 87, 07, 1d, 15, 19, 87, 07, 1d, 15, 19, 83, 07, 21, 87, 07, 1d, 15, 19, fb, 06, 8b, 07, 11, ff, 06, 83, 07, 21, 87, 07, 1d, 15, 19, 8f, 07, 9a, 07, 93, 07, 31, 97, 07, 2d, 25, 29, 9e, 07, 0d, a3, 07, 09, 05, 00, 62, 01, 03, 01, 02, 0c, 20, 05, 00, 02, 08, 00, 0c, 05, 00, 0d, 02, 06, 00, 02, 06, 00, 07, 03, 03, 09, 00, 0a, a3, 07, 00, 10, 00, 1d, 20, 09, 9e, 07, 00, 10, 00, 1d, 09, 01, 09, 01, 0a, 9e, 07, 02, 0f, 00, 1c, 20, 0d, 9a, 07, 00, 0f, 00, 1c, 0d, 01, 0c, 00, 19, 20, 35, 4a, 00, 0c, 00, 19, 4a, 00, 1d, 00, 2a, 20, 67, 46, 00, 1d, 00, 2a, 46, 00, 2e, 00, 3c, 20, 63, 41, 00, 2e, 00, 3c, 63, 00, 3d, 02, 0a, 41, 02, 0a, 00, 0b, 5f, 01, 09, 01, 12, 9a, 07, 03, 09, 00, 0f, 03, 03, 09, 01, 0c, 20, 45, 00, 01, 08, 00, 0c, 45, 00, 0d, 02, 06, 00, 02, 06, 00, 07, eb, 01, 02, 08, 00, 15, 20, 49, e6, 01, 00, 08, 00, 15, 49, 00, 16, 02, 06, e6, 01, 02, 0f, 00, 1c, 20, e2, 01, 31, 00, 0f, 00, 1c, e2, 01, 01, 0c, 00, 19, 20, 4d, de, 01, 00, 0c, 00, 19, de, 01, 00, 1d, 00, 2a, 20, 97, 02, da, 01, 00, 1d, 00, 2a, da, 01, 00, 2e, 00, 3c, 20, 93, 02, 59, 00, 2e, 00, 3c, 93, 02, 00, 3d, 02, 0a, 59, 02, 0a, 00, 0b, 8f, 02, 01, 09, 00, 17, 31, 02, 09, 00, 0f, 8b, 02, 03, 08, 00, 0c, 20, 5d, 00, 00, 08, 00, 0c, 5d, 01, 0d, 01, 10, 20, 61, 00, 01, 0c, 00, 10, 61, 00, 11, 02, 0a, 00, 02, 0a, 00, 0b, 8f, 03, 02, 0c, 00, 19, 20, 65, 8a, 03, 00, 0c, 00, 19, 65, 00, 1a, 02, 0a, 8a, 03, 03, 11, 00, 1e, 20, 86, 03, 2d, 00, 11, 00, 1e, 86, 03, 01, 10, 00, 1d, 20, 69, 82, 03, 00, 10, 00, 1d, 82, 03, 00, 21, 00, 2e, 20, bf, 03, fe, 02, 00, 21, 00, 2e, fe, 02, 00, 32, 00, 40, 20, bb, 03, 75, 00, 32, 00, 40, bb, 03, 00, 41, 02, 0e, 75, 02, 0e, 00, 0f, b7, 03, 01, 0d, 00, 1b, 2d, 02, 0d, 00, 13, 00, 02, 06, 00, 07, af, 03, 02, 09, 01, 0c, 20, 79, 00, 01, 08, 00, 0c, 79, 00, 0d, 02, 06, 00, 02, 06, 00, 07, cb, 06, 02, 09, 00, 0a, c7, 04, 00, 10, 00, 1d, 20, 7d, c2, 04, 00, 10, 00, 1d, 7d, 00, 1e, 02, 06, c2, 04, 02, 0f, 00, 1c, 20, be, 04, 29, 00, 0f, 00, 1c, be, 04, 01, 0c, 00, 19, 20, 81, 01, ba, 04, 00, 0c, 00, 19, ba, 04, 00, 1d, 00, 2a, 20, d7, 06, b6, 04, 00, 1d, 00, 2a, b6, 04, 00, 2e, 00, 3c, 20, d3, 06, 8d, 01, 00, 2e, 00, 3c, d3, 06, 00, 3d, 02, 0a, 8d, 01, 02, 0a, 00, 0b, cf, 06, 01, 09, 00, 17, 29, 02, 0d, 02, 0f, fb, 06, 05, 09, 00, 0a, cb, 06, 00, 10, 00, 1d, 20, 11, c6, 06, 00, 10, 00, 1d, 11, 00, 1e, 02, 06, c6, 06, 02, 0f, 00, 1c, 20, c2, 06, 25, 00, 0f, 00, 1c, c2, 06, 01, 0c, 00, 19, 20, 15, be, 06, 00, 0c, 00, 19, be, 06, 00, 1d, 00, 2a, 20, 87, 07, ba, 06, 00, 1d, 00, 2a, ba, 06, 00, 2e, 00, 3c, 20, 83, 07, 21, 00, 2e, 00, 3c, 83, 07, 00, 3d, 02, 0a, 21, 02, 0a, 00, 0b, ff, 06, 01, 09, 00, 17, 25, 02, 09, 00, 0f, f7, 06, 02, 01, 00, 02]
+ Raw bytes (1252): 0x[01, 01, da, 01, 09, 53, 57, 41, 5b, 3d, 35, 39, 05, 00, e7, 06, 09, 05, 00, e7, 06, 09, 05, 00, e2, 06, 0d, e7, 06, 09, 05, 00, 0d, 35, 0d, 35, 46, 39, 0d, 35, 46, 39, 0d, 35, 5b, 3d, 35, 39, 57, 41, 5b, 3d, 35, 39, e2, 06, 0d, e7, 06, 09, 05, 00, 45, 00, db, 01, 49, 45, 00, db, 01, 49, 45, 00, d6, 01, 31, db, 01, 49, 45, 00, d6, 01, 31, db, 01, 49, 45, 00, d2, 01, 4d, d6, 01, 31, db, 01, 49, 45, 00, d2, 01, 4d, d6, 01, 31, db, 01, 49, 45, 00, ce, 01, 51, d2, 01, 4d, d6, 01, 31, db, 01, 49, 45, 00, ce, 01, 51, d2, 01, 4d, d6, 01, 31, db, 01, 49, 45, 00, ff, 01, 55, 4d, 51, fb, 01, 59, ff, 01, 55, 4d, 51, 49, f7, 01, fb, 01, 59, ff, 01, 55, 4d, 51, 61, 00, f3, 02, 65, 61, 00, f3, 02, 65, 61, 00, ee, 02, 2d, f3, 02, 65, 61, 00, ee, 02, 2d, f3, 02, 65, 61, 00, ea, 02, 69, ee, 02, 2d, f3, 02, 65, 61, 00, ea, 02, 69, ee, 02, 2d, f3, 02, 65, 61, 00, e6, 02, 6d, ea, 02, 69, ee, 02, 2d, f3, 02, 65, 61, 00, e6, 02, 6d, ea, 02, 69, ee, 02, 2d, f3, 02, 65, 61, 00, 9b, 03, 71, 69, 6d, 97, 03, 75, 9b, 03, 71, 69, 6d, 8f, 03, 00, 65, 93, 03, 97, 03, 75, 9b, 03, 71, 69, 6d, 7d, 9b, 06, 9f, 06, 8d, 01, a3, 06, 89, 01, 81, 01, 85, 01, 79, 00, 9f, 04, 7d, 79, 00, 9f, 04, 7d, 79, 00, 9a, 04, 29, 9f, 04, 7d, 79, 00, 9a, 04, 29, 9f, 04, 7d, 79, 00, 96, 04, 81, 01, 9a, 04, 29, 9f, 04, 7d, 79, 00, 96, 04, 81, 01, 9a, 04, 29, 9f, 04, 7d, 79, 00, 92, 04, 85, 01, 96, 04, 81, 01, 9a, 04, 29, 9f, 04, 7d, 79, 00, 92, 04, 85, 01, 96, 04, 81, 01, 9a, 04, 29, 9f, 04, 7d, 79, 00, a3, 06, 89, 01, 81, 01, 85, 01, 9f, 06, 8d, 01, a3, 06, 89, 01, 81, 01, 85, 01, 11, c3, 06, c7, 06, 21, cb, 06, 1d, 15, 19, 7d, 9b, 06, 9f, 06, 8d, 01, a3, 06, 89, 01, 81, 01, 85, 01, 97, 06, 11, 7d, 9b, 06, 9f, 06, 8d, 01, a3, 06, 89, 01, 81, 01, 85, 01, 97, 06, 11, 7d, 9b, 06, 9f, 06, 8d, 01, a3, 06, 89, 01, 81, 01, 85, 01, 92, 06, 25, 97, 06, 11, 7d, 9b, 06, 9f, 06, 8d, 01, a3, 06, 89, 01, 81, 01, 85, 01, 92, 06, 25, 97, 06, 11, 7d, 9b, 06, 9f, 06, 8d, 01, a3, 06, 89, 01, 81, 01, 85, 01, 8e, 06, 15, 92, 06, 25, 97, 06, 11, 7d, 9b, 06, 9f, 06, 8d, 01, a3, 06, 89, 01, 81, 01, 85, 01, 8e, 06, 15, 92, 06, 25, 97, 06, 11, 7d, 9b, 06, 9f, 06, 8d, 01, a3, 06, 89, 01, 81, 01, 85, 01, 8a, 06, 19, 8e, 06, 15, 92, 06, 25, 97, 06, 11, 7d, 9b, 06, 9f, 06, 8d, 01, a3, 06, 89, 01, 81, 01, 85, 01, 8a, 06, 19, 8e, 06, 15, 92, 06, 25, 97, 06, 11, 7d, 9b, 06, 9f, 06, 8d, 01, a3, 06, 89, 01, 81, 01, 85, 01, cb, 06, 1d, 15, 19, c7, 06, 21, cb, 06, 1d, 15, 19, bf, 06, cf, 06, 11, c3, 06, c7, 06, 21, cb, 06, 1d, 15, 19, d3, 06, de, 06, d7, 06, 31, db, 06, 2d, 25, 29, e2, 06, 0d, e7, 06, 09, 05, 00, 62, 01, 03, 01, 02, 0c, 20, 05, 00, 02, 08, 00, 0c, 05, 00, 0d, 02, 06, 00, 02, 06, 00, 07, 03, 03, 09, 00, 0a, e7, 06, 00, 10, 00, 1d, 20, 09, e2, 06, 00, 10, 00, 1d, 09, 01, 09, 01, 0a, e2, 06, 02, 0f, 00, 1c, 20, 0d, de, 06, 00, 0f, 00, 1c, 0d, 01, 0c, 00, 19, 20, 35, 46, 00, 0c, 00, 19, 46, 00, 1d, 00, 2a, 20, 39, 42, 00, 1d, 00, 2a, 42, 00, 2e, 00, 3c, 20, 3d, 41, 00, 2e, 00, 3c, 57, 00, 3d, 02, 0a, 41, 02, 0a, 00, 0b, 53, 01, 09, 01, 12, de, 06, 03, 09, 00, 0f, 03, 03, 09, 01, 0c, 20, 45, 00, 01, 08, 00, 0c, 45, 00, 0d, 02, 06, 00, 02, 06, 00, 07, db, 01, 02, 08, 00, 15, 20, 49, d6, 01, 00, 08, 00, 15, 49, 00, 16, 02, 06, d6, 01, 02, 0f, 00, 1c, 20, d2, 01, 31, 00, 0f, 00, 1c, d2, 01, 01, 0c, 00, 19, 20, 4d, ce, 01, 00, 0c, 00, 19, ce, 01, 00, 1d, 00, 2a, 20, 51, ca, 01, 00, 1d, 00, 2a, ca, 01, 00, 2e, 00, 3c, 20, 55, 59, 00, 2e, 00, 3c, fb, 01, 00, 3d, 02, 0a, 59, 02, 0a, 00, 0b, f7, 01, 01, 09, 00, 17, 31, 02, 09, 00, 0f, f3, 01, 03, 08, 00, 0c, 20, 5d, 00, 00, 08, 00, 0c, 5d, 01, 0d, 01, 10, 20, 61, 00, 01, 0c, 00, 10, 61, 00, 11, 02, 0a, 00, 02, 0a, 00, 0b, f3, 02, 02, 0c, 00, 19, 20, 65, ee, 02, 00, 0c, 00, 19, 65, 00, 1a, 02, 0a, ee, 02, 03, 11, 00, 1e, 20, ea, 02, 2d, 00, 11, 00, 1e, ea, 02, 01, 10, 00, 1d, 20, 69, e6, 02, 00, 10, 00, 1d, e6, 02, 00, 21, 00, 2e, 20, 6d, e2, 02, 00, 21, 00, 2e, e2, 02, 00, 32, 00, 40, 20, 71, 75, 00, 32, 00, 40, 97, 03, 00, 41, 02, 0e, 75, 02, 0e, 00, 0f, 93, 03, 01, 0d, 00, 1b, 2d, 02, 0d, 00, 13, 00, 02, 06, 00, 07, 8b, 03, 02, 09, 01, 0c, 20, 79, 00, 01, 08, 00, 0c, 79, 00, 0d, 02, 06, 00, 02, 06, 00, 07, 97, 06, 02, 09, 00, 0a, 9f, 04, 00, 10, 00, 1d, 20, 7d, 9a, 04, 00, 10, 00, 1d, 7d, 00, 1e, 02, 06, 9a, 04, 02, 0f, 00, 1c, 20, 96, 04, 29, 00, 0f, 00, 1c, 96, 04, 01, 0c, 00, 19, 20, 81, 01, 92, 04, 00, 0c, 00, 19, 92, 04, 00, 1d, 00, 2a, 20, 85, 01, 8e, 04, 00, 1d, 00, 2a, 8e, 04, 00, 2e, 00, 3c, 20, 89, 01, 8d, 01, 00, 2e, 00, 3c, 9f, 06, 00, 3d, 02, 0a, 8d, 01, 02, 0a, 00, 0b, 9b, 06, 01, 09, 00, 17, 29, 02, 0d, 02, 0f, bf, 06, 05, 09, 00, 0a, 97, 06, 00, 10, 00, 1d, 20, 11, 92, 06, 00, 10, 00, 1d, 11, 00, 1e, 02, 06, 92, 06, 02, 0f, 00, 1c, 20, 8e, 06, 25, 00, 0f, 00, 1c, 8e, 06, 01, 0c, 00, 19, 20, 15, 8a, 06, 00, 0c, 00, 19, 8a, 06, 00, 1d, 00, 2a, 20, 19, 86, 06, 00, 1d, 00, 2a, 86, 06, 00, 2e, 00, 3c, 20, 1d, 21, 00, 2e, 00, 3c, c7, 06, 00, 3d, 02, 0a, 21, 02, 0a, 00, 0b, c3, 06, 01, 09, 00, 17, 25, 02, 09, 00, 0f, bb, 06, 02, 01, 00, 02]
4 - file 0 => global file 1
- Number of expressions: 233
- Number of expressions: 233
- - expression 0 operands: lhs = Counter(2), rhs = Expression(23, Add)
- - expression 1 operands: lhs = Expression(24, Add), rhs = Counter(16)
- - expression 2 operands: lhs = Expression(25, Add), rhs = Counter(15)
+ Number of expressions: 218
+ - expression 0 operands: lhs = Counter(2), rhs = Expression(20, Add)
+ - expression 1 operands: lhs = Expression(21, Add), rhs = Counter(16)
+ - expression 2 operands: lhs = Expression(22, Add), rhs = Counter(15)
9 - expression 3 operands: lhs = Counter(13), rhs = Counter(14)
10 - expression 4 operands: lhs = Counter(1), rhs = Zero
Build completed unsuccessfully in 0:13:07
- - expression 5 operands: lhs = Expression(232, Add), rhs = Counter(2)
+ - expression 5 operands: lhs = Expression(217, Add), rhs = Counter(2)
12 - expression 6 operands: lhs = Counter(1), rhs = Zero
- - expression 7 operands: lhs = Expression(232, Add), rhs = Counter(2)
+ - expression 7 operands: lhs = Expression(217, Add), rhs = Counter(2)
14 - expression 8 operands: lhs = Counter(1), rhs = Zero
- - expression 9 operands: lhs = Expression(231, Sub), rhs = Counter(3)
- - expression 10 operands: lhs = Expression(232, Add), rhs = Counter(2)
+ - expression 9 operands: lhs = Expression(216, Sub), rhs = Counter(3)
+ - expression 10 operands: lhs = Expression(217, Add), rhs = Counter(2)
17 - expression 11 operands: lhs = Counter(1), rhs = Zero
18 - expression 12 operands: lhs = Counter(3), rhs = Counter(13)
19 - expression 13 operands: lhs = Counter(3), rhs = Counter(13)
- - expression 14 operands: lhs = Counter(13), rhs = Counter(14)
- - expression 14 operands: lhs = Counter(13), rhs = Counter(14)
- - expression 15 operands: lhs = Expression(18, Sub), rhs = Counter(14)
- - expression 16 operands: lhs = Counter(3), rhs = Counter(13)
- - expression 17 operands: lhs = Expression(18, Sub), rhs = Counter(14)
- - expression 18 operands: lhs = Counter(3), rhs = Counter(13)
- - expression 19 operands: lhs = Expression(25, Add), rhs = Counter(15)
- - expression 20 operands: lhs = Counter(13), rhs = Counter(14)
- - expression 21 operands: lhs = Expression(25, Add), rhs = Counter(15)
+ - expression 14 operands: lhs = Expression(17, Sub), rhs = Counter(14)
+ - expression 15 operands: lhs = Counter(3), rhs = Counter(13)
+ - expression 16 operands: lhs = Expression(17, Sub), rhs = Counter(14)
+ - expression 17 operands: lhs = Counter(3), rhs = Counter(13)
+ - expression 18 operands: lhs = Expression(22, Add), rhs = Counter(15)
+ - expression 19 operands: lhs = Counter(13), rhs = Counter(14)
+ - expression 20 operands: lhs = Expression(21, Add), rhs = Counter(16)
+ - expression 21 operands: lhs = Expression(22, Add), rhs = Counter(15)
28 - expression 22 operands: lhs = Counter(13), rhs = Counter(14)
- - expression 23 operands: lhs = Expression(24, Add), rhs = Counter(16)
- - expression 24 operands: lhs = Expression(25, Add), rhs = Counter(15)
- - expression 25 operands: lhs = Counter(13), rhs = Counter(14)
- - expression 26 operands: lhs = Expression(231, Sub), rhs = Counter(3)
- - expression 27 operands: lhs = Expression(232, Add), rhs = Counter(2)
- - expression 28 operands: lhs = Counter(1), rhs = Zero
- - expression 29 operands: lhs = Counter(17), rhs = Zero
- - expression 30 operands: lhs = Expression(58, Add), rhs = Counter(18)
- - expression 31 operands: lhs = Counter(17), rhs = Zero
- - expression 32 operands: lhs = Expression(58, Add), rhs = Counter(18)
+ - expression 23 operands: lhs = Expression(216, Sub), rhs = Counter(3)
+ - expression 24 operands: lhs = Expression(217, Add), rhs = Counter(2)
+ - expression 25 operands: lhs = Counter(1), rhs = Zero
+ - expression 26 operands: lhs = Counter(17), rhs = Zero
+ - expression 27 operands: lhs = Expression(54, Add), rhs = Counter(18)
+ - expression 28 operands: lhs = Counter(17), rhs = Zero
+ - expression 29 operands: lhs = Expression(54, Add), rhs = Counter(18)
+ - expression 30 operands: lhs = Counter(17), rhs = Zero
+ - expression 31 operands: lhs = Expression(53, Sub), rhs = Counter(12)
+ - expression 32 operands: lhs = Expression(54, Add), rhs = Counter(18)
39 - expression 33 operands: lhs = Counter(17), rhs = Zero
- - expression 34 operands: lhs = Expression(57, Sub), rhs = Counter(12)
- - expression 35 operands: lhs = Expression(58, Add), rhs = Counter(18)
+ - expression 34 operands: lhs = Expression(53, Sub), rhs = Counter(12)
+ - expression 35 operands: lhs = Expression(54, Add), rhs = Counter(18)
42 - expression 36 operands: lhs = Counter(17), rhs = Zero
- - expression 37 operands: lhs = Expression(57, Sub), rhs = Counter(12)
- - expression 38 operands: lhs = Expression(58, Add), rhs = Counter(18)
- - expression 39 operands: lhs = Counter(17), rhs = Zero
- - expression 40 operands: lhs = Expression(56, Sub), rhs = Counter(19)
- - expression 41 operands: lhs = Expression(57, Sub), rhs = Counter(12)
- - expression 42 operands: lhs = Expression(58, Add), rhs = Counter(18)
- - expression 43 operands: lhs = Counter(17), rhs = Zero
- - expression 44 operands: lhs = Expression(56, Sub), rhs = Counter(19)
- - expression 45 operands: lhs = Expression(57, Sub), rhs = Counter(12)
- - expression 46 operands: lhs = Expression(58, Add), rhs = Counter(18)
- - expression 47 operands: lhs = Counter(17), rhs = Zero
- - expression 48 operands: lhs = Counter(19), rhs = Counter(20)
- - expression 49 operands: lhs = Expression(55, Sub), rhs = Counter(20)
- - expression 50 operands: lhs = Expression(56, Sub), rhs = Counter(19)
- - expression 51 operands: lhs = Expression(57, Sub), rhs = Counter(12)
- - expression 52 operands: lhs = Expression(58, Add), rhs = Counter(18)
- - expression 53 operands: lhs = Counter(17), rhs = Zero
- - expression 54 operands: lhs = Expression(55, Sub), rhs = Counter(20)
- - expression 55 operands: lhs = Expression(56, Sub), rhs = Counter(19)
- - expression 56 operands: lhs = Expression(57, Sub), rhs = Counter(12)
- - expression 57 operands: lhs = Expression(58, Add), rhs = Counter(18)
- - expression 58 operands: lhs = Counter(17), rhs = Zero
- - expression 59 operands: lhs = Expression(69, Add), rhs = Counter(21)
- - expression 60 operands: lhs = Counter(19), rhs = Counter(20)
- - expression 61 operands: lhs = Expression(69, Add), rhs = Counter(21)
- - expression 62 operands: lhs = Counter(19), rhs = Counter(20)
- - expression 63 operands: lhs = Expression(68, Add), rhs = Counter(22)
- - expression 64 operands: lhs = Expression(69, Add), rhs = Counter(21)
- - expression 65 operands: lhs = Counter(19), rhs = Counter(20)
- - expression 66 operands: lhs = Counter(18), rhs = Expression(67, Add)
- - expression 67 operands: lhs = Expression(68, Add), rhs = Counter(22)
- - expression 68 operands: lhs = Expression(69, Add), rhs = Counter(21)
- - expression 69 operands: lhs = Counter(19), rhs = Counter(20)
- - expression 70 operands: lhs = Counter(24), rhs = Zero
- - expression 71 operands: lhs = Expression(99, Add), rhs = Counter(25)
- - expression 72 operands: lhs = Counter(24), rhs = Zero
- - expression 73 operands: lhs = Expression(99, Add), rhs = Counter(25)
+ - expression 37 operands: lhs = Expression(52, Sub), rhs = Counter(19)
+ - expression 38 operands: lhs = Expression(53, Sub), rhs = Counter(12)
+ - expression 39 operands: lhs = Expression(54, Add), rhs = Counter(18)
+ - expression 40 operands: lhs = Counter(17), rhs = Zero
+ - expression 41 operands: lhs = Expression(52, Sub), rhs = Counter(19)
+ - expression 42 operands: lhs = Expression(53, Sub), rhs = Counter(12)
+ - expression 43 operands: lhs = Expression(54, Add), rhs = Counter(18)
+ - expression 44 operands: lhs = Counter(17), rhs = Zero
+ - expression 45 operands: lhs = Expression(51, Sub), rhs = Counter(20)
+ - expression 46 operands: lhs = Expression(52, Sub), rhs = Counter(19)
+ - expression 47 operands: lhs = Expression(53, Sub), rhs = Counter(12)
+ - expression 48 operands: lhs = Expression(54, Add), rhs = Counter(18)
+ - expression 49 operands: lhs = Counter(17), rhs = Zero
+ - expression 50 operands: lhs = Expression(51, Sub), rhs = Counter(20)
+ - expression 51 operands: lhs = Expression(52, Sub), rhs = Counter(19)
+ - expression 52 operands: lhs = Expression(53, Sub), rhs = Counter(12)
+ - expression 53 operands: lhs = Expression(54, Add), rhs = Counter(18)
+ - expression 54 operands: lhs = Counter(17), rhs = Zero
+ - expression 55 operands: lhs = Expression(63, Add), rhs = Counter(21)
+ - expression 56 operands: lhs = Counter(19), rhs = Counter(20)
+ - expression 57 operands: lhs = Expression(62, Add), rhs = Counter(22)
+ - expression 58 operands: lhs = Expression(63, Add), rhs = Counter(21)
+ - expression 59 operands: lhs = Counter(19), rhs = Counter(20)
+ - expression 60 operands: lhs = Counter(18), rhs = Expression(61, Add)
+ - expression 61 operands: lhs = Expression(62, Add), rhs = Counter(22)
+ - expression 62 operands: lhs = Expression(63, Add), rhs = Counter(21)
+ - expression 63 operands: lhs = Counter(19), rhs = Counter(20)
+ - expression 64 operands: lhs = Counter(24), rhs = Zero
+ - expression 65 operands: lhs = Expression(92, Add), rhs = Counter(25)
+ - expression 66 operands: lhs = Counter(24), rhs = Zero
+ - expression 67 operands: lhs = Expression(92, Add), rhs = Counter(25)
+ - expression 68 operands: lhs = Counter(24), rhs = Zero
+ - expression 69 operands: lhs = Expression(91, Sub), rhs = Counter(11)
+ - expression 70 operands: lhs = Expression(92, Add), rhs = Counter(25)
+ - expression 71 operands: lhs = Counter(24), rhs = Zero
+ - expression 72 operands: lhs = Expression(91, Sub), rhs = Counter(11)
+ - expression 73 operands: lhs = Expression(92, Add), rhs = Counter(25)
80 - expression 74 operands: lhs = Counter(24), rhs = Zero
- - expression 75 operands: lhs = Expression(98, Sub), rhs = Counter(11)
- - expression 76 operands: lhs = Expression(99, Add), rhs = Counter(25)
- - expression 77 operands: lhs = Counter(24), rhs = Zero
- - expression 78 operands: lhs = Expression(98, Sub), rhs = Counter(11)
- - expression 79 operands: lhs = Expression(99, Add), rhs = Counter(25)
- - expression 80 operands: lhs = Counter(24), rhs = Zero
- - expression 81 operands: lhs = Expression(97, Sub), rhs = Counter(26)
- - expression 82 operands: lhs = Expression(98, Sub), rhs = Counter(11)
- - expression 83 operands: lhs = Expression(99, Add), rhs = Counter(25)
- - expression 84 operands: lhs = Counter(24), rhs = Zero
- - expression 85 operands: lhs = Expression(97, Sub), rhs = Counter(26)
- - expression 86 operands: lhs = Expression(98, Sub), rhs = Counter(11)
- - expression 87 operands: lhs = Expression(99, Add), rhs = Counter(25)
- - expression 88 operands: lhs = Counter(24), rhs = Zero
- - expression 89 operands: lhs = Counter(26), rhs = Counter(27)
- - expression 90 operands: lhs = Expression(96, Sub), rhs = Counter(27)
- - expression 91 operands: lhs = Expression(97, Sub), rhs = Counter(26)
- - expression 92 operands: lhs = Expression(98, Sub), rhs = Counter(11)
- - expression 93 operands: lhs = Expression(99, Add), rhs = Counter(25)
- - expression 94 operands: lhs = Counter(24), rhs = Zero
- - expression 95 operands: lhs = Expression(96, Sub), rhs = Counter(27)
- - expression 96 operands: lhs = Expression(97, Sub), rhs = Counter(26)
- - expression 97 operands: lhs = Expression(98, Sub), rhs = Counter(11)
- - expression 98 operands: lhs = Expression(99, Add), rhs = Counter(25)
- - expression 99 operands: lhs = Counter(24), rhs = Zero
- - expression 100 operands: lhs = Expression(111, Add), rhs = Counter(28)
- - expression 101 operands: lhs = Counter(26), rhs = Counter(27)
- - expression 102 operands: lhs = Expression(111, Add), rhs = Counter(28)
- - expression 103 operands: lhs = Counter(26), rhs = Counter(27)
- - expression 104 operands: lhs = Expression(110, Add), rhs = Counter(29)
- - expression 105 operands: lhs = Expression(111, Add), rhs = Counter(28)
- - expression 106 operands: lhs = Counter(26), rhs = Counter(27)
- - expression 107 operands: lhs = Expression(108, Add), rhs = Zero
- - expression 108 operands: lhs = Counter(25), rhs = Expression(109, Add)
- - expression 109 operands: lhs = Expression(110, Add), rhs = Counter(29)
- - expression 110 operands: lhs = Expression(111, Add), rhs = Counter(28)
- - expression 111 operands: lhs = Counter(26), rhs = Counter(27)
- - expression 112 operands: lhs = Counter(31), rhs = Expression(211, Add)
- - expression 113 operands: lhs = Expression(212, Add), rhs = Counter(35)
- - expression 114 operands: lhs = Expression(213, Add), rhs = Counter(34)
- - expression 115 operands: lhs = Counter(32), rhs = Counter(33)
- - expression 116 operands: lhs = Counter(30), rhs = Zero
- - expression 117 operands: lhs = Expression(145, Add), rhs = Counter(31)
- - expression 118 operands: lhs = Counter(30), rhs = Zero
- - expression 119 operands: lhs = Expression(145, Add), rhs = Counter(31)
- - expression 120 operands: lhs = Counter(30), rhs = Zero
- - expression 121 operands: lhs = Expression(144, Sub), rhs = Counter(10)
- - expression 122 operands: lhs = Expression(145, Add), rhs = Counter(31)
- - expression 123 operands: lhs = Counter(30), rhs = Zero
- - expression 124 operands: lhs = Expression(144, Sub), rhs = Counter(10)
- - expression 125 operands: lhs = Expression(145, Add), rhs = Counter(31)
- - expression 126 operands: lhs = Counter(30), rhs = Zero
- - expression 127 operands: lhs = Expression(143, Sub), rhs = Counter(32)
- - expression 128 operands: lhs = Expression(144, Sub), rhs = Counter(10)
- - expression 129 operands: lhs = Expression(145, Add), rhs = Counter(31)
+ - expression 75 operands: lhs = Expression(90, Sub), rhs = Counter(26)
+ - expression 76 operands: lhs = Expression(91, Sub), rhs = Counter(11)
+ - expression 77 operands: lhs = Expression(92, Add), rhs = Counter(25)
+ - expression 78 operands: lhs = Counter(24), rhs = Zero
+ - expression 79 operands: lhs = Expression(90, Sub), rhs = Counter(26)
+ - expression 80 operands: lhs = Expression(91, Sub), rhs = Counter(11)
+ - expression 81 operands: lhs = Expression(92, Add), rhs = Counter(25)
+ - expression 82 operands: lhs = Counter(24), rhs = Zero
+ - expression 83 operands: lhs = Expression(89, Sub), rhs = Counter(27)
+ - expression 84 operands: lhs = Expression(90, Sub), rhs = Counter(26)
+ - expression 85 operands: lhs = Expression(91, Sub), rhs = Counter(11)
+ - expression 86 operands: lhs = Expression(92, Add), rhs = Counter(25)
+ - expression 87 operands: lhs = Counter(24), rhs = Zero
+ - expression 88 operands: lhs = Expression(89, Sub), rhs = Counter(27)
+ - expression 89 operands: lhs = Expression(90, Sub), rhs = Counter(26)
+ - expression 90 operands: lhs = Expression(91, Sub), rhs = Counter(11)
+ - expression 91 operands: lhs = Expression(92, Add), rhs = Counter(25)
+ - expression 92 operands: lhs = Counter(24), rhs = Zero
+ - expression 93 operands: lhs = Expression(102, Add), rhs = Counter(28)
+ - expression 94 operands: lhs = Counter(26), rhs = Counter(27)
+ - expression 95 operands: lhs = Expression(101, Add), rhs = Counter(29)
+ - expression 96 operands: lhs = Expression(102, Add), rhs = Counter(28)
+ - expression 97 operands: lhs = Counter(26), rhs = Counter(27)
+ - expression 98 operands: lhs = Expression(99, Add), rhs = Zero
+ - expression 99 operands: lhs = Counter(25), rhs = Expression(100, Add)
+ - expression 100 operands: lhs = Expression(101, Add), rhs = Counter(29)
+ - expression 101 operands: lhs = Expression(102, Add), rhs = Counter(28)
+ - expression 102 operands: lhs = Counter(26), rhs = Counter(27)
+ - expression 103 operands: lhs = Counter(31), rhs = Expression(198, Add)
+ - expression 104 operands: lhs = Expression(199, Add), rhs = Counter(35)
+ - expression 105 operands: lhs = Expression(200, Add), rhs = Counter(34)
+ - expression 106 operands: lhs = Counter(32), rhs = Counter(33)
+ - expression 107 operands: lhs = Counter(30), rhs = Zero
+ - expression 108 operands: lhs = Expression(135, Add), rhs = Counter(31)
+ - expression 109 operands: lhs = Counter(30), rhs = Zero
+ - expression 110 operands: lhs = Expression(135, Add), rhs = Counter(31)
+ - expression 111 operands: lhs = Counter(30), rhs = Zero
+ - expression 112 operands: lhs = Expression(134, Sub), rhs = Counter(10)
+ - expression 113 operands: lhs = Expression(135, Add), rhs = Counter(31)
+ - expression 114 operands: lhs = Counter(30), rhs = Zero
+ - expression 115 operands: lhs = Expression(134, Sub), rhs = Counter(10)
+ - expression 116 operands: lhs = Expression(135, Add), rhs = Counter(31)
+ - expression 117 operands: lhs = Counter(30), rhs = Zero
+ - expression 118 operands: lhs = Expression(133, Sub), rhs = Counter(32)
+ - expression 119 operands: lhs = Expression(134, Sub), rhs = Counter(10)
+ - expression 120 operands: lhs = Expression(135, Add), rhs = Counter(31)
+ - expression 121 operands: lhs = Counter(30), rhs = Zero
+ - expression 122 operands: lhs = Expression(133, Sub), rhs = Counter(32)
+ - expression 123 operands: lhs = Expression(134, Sub), rhs = Counter(10)
+ - expression 124 operands: lhs = Expression(135, Add), rhs = Counter(31)
+ - expression 125 operands: lhs = Counter(30), rhs = Zero
+ - expression 126 operands: lhs = Expression(132, Sub), rhs = Counter(33)
+ - expression 127 operands: lhs = Expression(133, Sub), rhs = Counter(32)
+ - expression 128 operands: lhs = Expression(134, Sub), rhs = Counter(10)
+ - expression 129 operands: lhs = Expression(135, Add), rhs = Counter(31)
136 - expression 130 operands: lhs = Counter(30), rhs = Zero
- - expression 131 operands: lhs = Expression(143, Sub), rhs = Counter(32)
- - expression 132 operands: lhs = Expression(144, Sub), rhs = Counter(10)
- - expression 133 operands: lhs = Expression(145, Add), rhs = Counter(31)
- - expression 134 operands: lhs = Counter(30), rhs = Zero
- - expression 135 operands: lhs = Counter(32), rhs = Counter(33)
- - expression 136 operands: lhs = Expression(142, Sub), rhs = Counter(33)
- - expression 137 operands: lhs = Expression(143, Sub), rhs = Counter(32)
- - expression 138 operands: lhs = Expression(144, Sub), rhs = Counter(10)
- - expression 139 operands: lhs = Expression(145, Add), rhs = Counter(31)
- - expression 140 operands: lhs = Counter(30), rhs = Zero
- - expression 141 operands: lhs = Expression(142, Sub), rhs = Counter(33)
- - expression 142 operands: lhs = Expression(143, Sub), rhs = Counter(32)
- - expression 143 operands: lhs = Expression(144, Sub), rhs = Counter(10)
- - expression 144 operands: lhs = Expression(145, Add), rhs = Counter(31)
- - expression 145 operands: lhs = Counter(30), rhs = Zero
- - expression 146 operands: lhs = Expression(213, Add), rhs = Counter(34)
- - expression 147 operands: lhs = Counter(32), rhs = Counter(33)
- - expression 148 operands: lhs = Expression(213, Add), rhs = Counter(34)
- - expression 149 operands: lhs = Counter(32), rhs = Counter(33)
- - expression 150 operands: lhs = Expression(212, Add), rhs = Counter(35)
- - expression 151 operands: lhs = Expression(213, Add), rhs = Counter(34)
- - expression 152 operands: lhs = Counter(32), rhs = Counter(33)
- - expression 153 operands: lhs = Counter(4), rhs = Expression(223, Add)
- - expression 154 operands: lhs = Expression(224, Add), rhs = Counter(8)
- - expression 155 operands: lhs = Expression(225, Add), rhs = Counter(7)
- - expression 156 operands: lhs = Counter(5), rhs = Counter(6)
- - expression 157 operands: lhs = Counter(31), rhs = Expression(211, Add)
- - expression 158 operands: lhs = Expression(212, Add), rhs = Counter(35)
- - expression 159 operands: lhs = Expression(213, Add), rhs = Counter(34)
- - expression 160 operands: lhs = Counter(32), rhs = Counter(33)
- - expression 161 operands: lhs = Expression(210, Add), rhs = Counter(4)
- - expression 162 operands: lhs = Counter(31), rhs = Expression(211, Add)
- - expression 163 operands: lhs = Expression(212, Add), rhs = Counter(35)
- - expression 164 operands: lhs = Expression(213, Add), rhs = Counter(34)
- - expression 165 operands: lhs = Counter(32), rhs = Counter(33)
- - expression 166 operands: lhs = Expression(210, Add), rhs = Counter(4)
- - expression 167 operands: lhs = Counter(31), rhs = Expression(211, Add)
- - expression 168 operands: lhs = Expression(212, Add), rhs = Counter(35)
- - expression 169 operands: lhs = Expression(213, Add), rhs = Counter(34)
+ - expression 131 operands: lhs = Expression(132, Sub), rhs = Counter(33)
+ - expression 132 operands: lhs = Expression(133, Sub), rhs = Counter(32)
+ - expression 133 operands: lhs = Expression(134, Sub), rhs = Counter(10)
+ - expression 134 operands: lhs = Expression(135, Add), rhs = Counter(31)
+ - expression 135 operands: lhs = Counter(30), rhs = Zero
+ - expression 136 operands: lhs = Expression(200, Add), rhs = Counter(34)
+ - expression 137 operands: lhs = Counter(32), rhs = Counter(33)
+ - expression 138 operands: lhs = Expression(199, Add), rhs = Counter(35)
+ - expression 139 operands: lhs = Expression(200, Add), rhs = Counter(34)
+ - expression 140 operands: lhs = Counter(32), rhs = Counter(33)
+ - expression 141 operands: lhs = Counter(4), rhs = Expression(208, Add)
+ - expression 142 operands: lhs = Expression(209, Add), rhs = Counter(8)
+ - expression 143 operands: lhs = Expression(210, Add), rhs = Counter(7)
+ - expression 144 operands: lhs = Counter(5), rhs = Counter(6)
+ - expression 145 operands: lhs = Counter(31), rhs = Expression(198, Add)
+ - expression 146 operands: lhs = Expression(199, Add), rhs = Counter(35)
+ - expression 147 operands: lhs = Expression(200, Add), rhs = Counter(34)
+ - expression 148 operands: lhs = Counter(32), rhs = Counter(33)
+ - expression 149 operands: lhs = Expression(197, Add), rhs = Counter(4)
+ - expression 150 operands: lhs = Counter(31), rhs = Expression(198, Add)
+ - expression 151 operands: lhs = Expression(199, Add), rhs = Counter(35)
+ - expression 152 operands: lhs = Expression(200, Add), rhs = Counter(34)
+ - expression 153 operands: lhs = Counter(32), rhs = Counter(33)
+ - expression 154 operands: lhs = Expression(197, Add), rhs = Counter(4)
+ - expression 155 operands: lhs = Counter(31), rhs = Expression(198, Add)
+ - expression 156 operands: lhs = Expression(199, Add), rhs = Counter(35)
+ - expression 157 operands: lhs = Expression(200, Add), rhs = Counter(34)
+ - expression 158 operands: lhs = Counter(32), rhs = Counter(33)
+ - expression 159 operands: lhs = Expression(196, Sub), rhs = Counter(9)
+ - expression 160 operands: lhs = Expression(197, Add), rhs = Counter(4)
+ - expression 161 operands: lhs = Counter(31), rhs = Expression(198, Add)
+ - expression 162 operands: lhs = Expression(199, Add), rhs = Counter(35)
+ - expression 163 operands: lhs = Expression(200, Add), rhs = Counter(34)
+ - expression 164 operands: lhs = Counter(32), rhs = Counter(33)
+ - expression 165 operands: lhs = Expression(196, Sub), rhs = Counter(9)
+ - expression 166 operands: lhs = Expression(197, Add), rhs = Counter(4)
+ - expression 167 operands: lhs = Counter(31), rhs = Expression(198, Add)
+ - expression 168 operands: lhs = Expression(199, Add), rhs = Counter(35)
+ - expression 169 operands: lhs = Expression(200, Add), rhs = Counter(34)
176 - expression 170 operands: lhs = Counter(32), rhs = Counter(33)
- - expression 171 operands: lhs = Expression(209, Sub), rhs = Counter(9)
- - expression 172 operands: lhs = Expression(210, Add), rhs = Counter(4)
- - expression 173 operands: lhs = Counter(31), rhs = Expression(211, Add)
- - expression 174 operands: lhs = Expression(212, Add), rhs = Counter(35)
- - expression 175 operands: lhs = Expression(213, Add), rhs = Counter(34)
- - expression 176 operands: lhs = Counter(32), rhs = Counter(33)
- - expression 177 operands: lhs = Expression(209, Sub), rhs = Counter(9)
- - expression 178 operands: lhs = Expression(210, Add), rhs = Counter(4)
- - expression 179 operands: lhs = Counter(31), rhs = Expression(211, Add)
- - expression 180 operands: lhs = Expression(212, Add), rhs = Counter(35)
- - expression 181 operands: lhs = Expression(213, Add), rhs = Counter(34)
- - expression 182 operands: lhs = Counter(32), rhs = Counter(33)
- - expression 183 operands: lhs = Expression(208, Sub), rhs = Counter(5)
- - expression 184 operands: lhs = Expression(209, Sub), rhs = Counter(9)
- - expression 185 operands: lhs = Expression(210, Add), rhs = Counter(4)
- - expression 186 operands: lhs = Counter(31), rhs = Expression(211, Add)
- - expression 187 operands: lhs = Expression(212, Add), rhs = Counter(35)
- - expression 188 operands: lhs = Expression(213, Add), rhs = Counter(34)
- - expression 189 operands: lhs = Counter(32), rhs = Counter(33)
- - expression 190 operands: lhs = Expression(208, Sub), rhs = Counter(5)
- - expression 191 operands: lhs = Expression(209, Sub), rhs = Counter(9)
- - expression 192 operands: lhs = Expression(210, Add), rhs = Counter(4)
- - expression 193 operands: lhs = Counter(31), rhs = Expression(211, Add)
- - expression 194 operands: lhs = Expression(212, Add), rhs = Counter(35)
- - expression 195 operands: lhs = Expression(213, Add), rhs = Counter(34)
- - expression 196 operands: lhs = Counter(32), rhs = Counter(33)
- - expression 197 operands: lhs = Counter(5), rhs = Counter(6)
- - expression 198 operands: lhs = Expression(207, Sub), rhs = Counter(6)
- - expression 199 operands: lhs = Expression(208, Sub), rhs = Counter(5)
- - expression 200 operands: lhs = Expression(209, Sub), rhs = Counter(9)
- - expression 201 operands: lhs = Expression(210, Add), rhs = Counter(4)
- - expression 202 operands: lhs = Counter(31), rhs = Expression(211, Add)
- - expression 203 operands: lhs = Expression(212, Add), rhs = Counter(35)
- - expression 204 operands: lhs = Expression(213, Add), rhs = Counter(34)
- - expression 205 operands: lhs = Counter(32), rhs = Counter(33)
- - expression 206 operands: lhs = Expression(207, Sub), rhs = Counter(6)
- - expression 207 operands: lhs = Expression(208, Sub), rhs = Counter(5)
- - expression 208 operands: lhs = Expression(209, Sub), rhs = Counter(9)
- - expression 209 operands: lhs = Expression(210, Add), rhs = Counter(4)
- - expression 210 operands: lhs = Counter(31), rhs = Expression(211, Add)
- - expression 211 operands: lhs = Expression(212, Add), rhs = Counter(35)
- - expression 212 operands: lhs = Expression(213, Add), rhs = Counter(34)
- - expression 213 operands: lhs = Counter(32), rhs = Counter(33)
- - expression 214 operands: lhs = Expression(225, Add), rhs = Counter(7)
- - expression 215 operands: lhs = Counter(5), rhs = Counter(6)
- - expression 216 operands: lhs = Expression(225, Add), rhs = Counter(7)
- - expression 217 operands: lhs = Counter(5), rhs = Counter(6)
- - expression 218 operands: lhs = Expression(224, Add), rhs = Counter(8)
- - expression 219 operands: lhs = Expression(225, Add), rhs = Counter(7)
- - expression 220 operands: lhs = Counter(5), rhs = Counter(6)
- - expression 221 operands: lhs = Expression(222, Add), rhs = Expression(226, Add)
- - expression 222 operands: lhs = Counter(4), rhs = Expression(223, Add)
- - expression 223 operands: lhs = Expression(224, Add), rhs = Counter(8)
- - expression 224 operands: lhs = Expression(225, Add), rhs = Counter(7)
- - expression 225 operands: lhs = Counter(5), rhs = Counter(6)
- - expression 226 operands: lhs = Expression(227, Add), rhs = Expression(230, Sub)
- - expression 227 operands: lhs = Expression(228, Add), rhs = Counter(12)
- - expression 228 operands: lhs = Expression(229, Add), rhs = Counter(11)
- - expression 229 operands: lhs = Counter(9), rhs = Counter(10)
- - expression 230 operands: lhs = Expression(231, Sub), rhs = Counter(3)
- - expression 231 operands: lhs = Expression(232, Add), rhs = Counter(2)
- - expression 232 operands: lhs = Counter(1), rhs = Zero
+ - expression 171 operands: lhs = Expression(195, Sub), rhs = Counter(5)
+ - expression 172 operands: lhs = Expression(196, Sub), rhs = Counter(9)
+ - expression 173 operands: lhs = Expression(197, Add), rhs = Counter(4)
+ - expression 174 operands: lhs = Counter(31), rhs = Expression(198, Add)
+ - expression 175 operands: lhs = Expression(199, Add), rhs = Counter(35)
+ - expression 176 operands: lhs = Expression(200, Add), rhs = Counter(34)
+ - expression 177 operands: lhs = Counter(32), rhs = Counter(33)
+ - expression 178 operands: lhs = Expression(195, Sub), rhs = Counter(5)
+ - expression 179 operands: lhs = Expression(196, Sub), rhs = Counter(9)
+ - expression 180 operands: lhs = Expression(197, Add), rhs = Counter(4)
+ - expression 181 operands: lhs = Counter(31), rhs = Expression(198, Add)
+ - expression 182 operands: lhs = Expression(199, Add), rhs = Counter(35)
+ - expression 183 operands: lhs = Expression(200, Add), rhs = Counter(34)
+ - expression 184 operands: lhs = Counter(32), rhs = Counter(33)
+ - expression 185 operands: lhs = Expression(194, Sub), rhs = Counter(6)
+ - expression 186 operands: lhs = Expression(195, Sub), rhs = Counter(5)
+ - expression 187 operands: lhs = Expression(196, Sub), rhs = Counter(9)
+ - expression 188 operands: lhs = Expression(197, Add), rhs = Counter(4)
+ - expression 189 operands: lhs = Counter(31), rhs = Expression(198, Add)
+ - expression 190 operands: lhs = Expression(199, Add), rhs = Counter(35)
+ - expression 191 operands: lhs = Expression(200, Add), rhs = Counter(34)
+ - expression 192 operands: lhs = Counter(32), rhs = Counter(33)
+ - expression 193 operands: lhs = Expression(194, Sub), rhs = Counter(6)
+ - expression 194 operands: lhs = Expression(195, Sub), rhs = Counter(5)
+ - expression 195 operands: lhs = Expression(196, Sub), rhs = Counter(9)
+ - expression 196 operands: lhs = Expression(197, Add), rhs = Counter(4)
+ - expression 197 operands: lhs = Counter(31), rhs = Expression(198, Add)
+ - expression 198 operands: lhs = Expression(199, Add), rhs = Counter(35)
+ - expression 199 operands: lhs = Expression(200, Add), rhs = Counter(34)
+ - expression 200 operands: lhs = Counter(32), rhs = Counter(33)
+ - expression 201 operands: lhs = Expression(210, Add), rhs = Counter(7)
+ - expression 202 operands: lhs = Counter(5), rhs = Counter(6)
+ - expression 203 operands: lhs = Expression(209, Add), rhs = Counter(8)
+ - expression 204 operands: lhs = Expression(210, Add), rhs = Counter(7)
+ - expression 205 operands: lhs = Counter(5), rhs = Counter(6)
+ - expression 206 operands: lhs = Expression(207, Add), rhs = Expression(211, Add)
+ - expression 207 operands: lhs = Counter(4), rhs = Expression(208, Add)
+ - expression 208 operands: lhs = Expression(209, Add), rhs = Counter(8)
+ - expression 209 operands: lhs = Expression(210, Add), rhs = Counter(7)
+ - expression 210 operands: lhs = Counter(5), rhs = Counter(6)
+ - expression 211 operands: lhs = Expression(212, Add), rhs = Expression(215, Sub)
+ - expression 212 operands: lhs = Expression(213, Add), rhs = Counter(12)
+ - expression 213 operands: lhs = Expression(214, Add), rhs = Counter(11)
+ - expression 214 operands: lhs = Counter(9), rhs = Counter(10)
+ - expression 215 operands: lhs = Expression(216, Sub), rhs = Counter(3)
+ - expression 216 operands: lhs = Expression(217, Add), rhs = Counter(2)
+ - expression 217 operands: lhs = Counter(1), rhs = Zero
239 Number of file 0 mappings: 98
240 - Code(Counter(0)) at (prev + 3, 1) to (start + 2, 12)
241 - Branch { true: Counter(1), false: Zero } at (prev + 2, 8) to (start + 0, 12)

245 - Code(Zero) at (prev + 2, 6) to (start + 0, 7)
246 - Code(Expression(0, Add)) at (prev + 3, 9) to (start + 0, 10)
247     = (c2 + (((c13 + c14) + c15) + c16))
- - Code(Expression(232, Add)) at (prev + 0, 16) to (start + 0, 29)
+ - Code(Expression(217, Add)) at (prev + 0, 16) to (start + 0, 29)
249     = (c1 + Zero)
- - Branch { true: Counter(2), false: Expression(231, Sub) } at (prev + 0, 16) to (start + 0, 29)
+ - Branch { true: Counter(2), false: Expression(216, Sub) } at (prev + 0, 16) to (start + 0, 29)
251     true  = c2
252     false = ((c1 + Zero) - c2)
253 - Code(Counter(2)) at (prev + 1, 9) to (start + 1, 10)

- - Code(Expression(231, Sub)) at (prev + 2, 15) to (start + 0, 28)
+ - Code(Expression(216, Sub)) at (prev + 2, 15) to (start + 0, 28)
255     = ((c1 + Zero) - c2)
- - Branch { true: Counter(3), false: Expression(230, Sub) } at (prev + 0, 15) to (start + 0, 28)
+ - Branch { true: Counter(3), false: Expression(215, Sub) } at (prev + 0, 15) to (start + 0, 28)
257     true  = c3
258     false = (((c1 + Zero) - c2) - c3)
259 - Code(Counter(3)) at (prev + 1, 12) to (start + 0, 25)

- - Branch { true: Counter(13), false: Expression(18, Sub) } at (prev + 0, 12) to (start + 0, 25)
+ - Branch { true: Counter(13), false: Expression(17, Sub) } at (prev + 0, 12) to (start + 0, 25)
261     true  = c13
262     false = (c3 - c13)
- - Code(Expression(18, Sub)) at (prev + 0, 29) to (start + 0, 42)
+ - Code(Expression(17, Sub)) at (prev + 0, 29) to (start + 0, 42)
264     = (c3 - c13)
- - Branch { true: Expression(25, Add), false: Expression(17, Sub) } at (prev + 0, 29) to (start + 0, 42)
-     true  = (c13 + c14)
+ - Branch { true: Counter(14), false: Expression(16, Sub) } at (prev + 0, 29) to (start + 0, 42)
+     true  = c14
267     false = ((c3 - c13) - c14)
- - Code(Expression(17, Sub)) at (prev + 0, 46) to (start + 0, 60)
+ - Code(Expression(16, Sub)) at (prev + 0, 46) to (start + 0, 60)
269     = ((c3 - c13) - c14)
- - Branch { true: Expression(24, Add), false: Counter(16) } at (prev + 0, 46) to (start + 0, 60)
-     true  = ((c13 + c14) + c15)
+ - Branch { true: Counter(15), false: Counter(16) } at (prev + 0, 46) to (start + 0, 60)
+     true  = c15
272     false = c16
- - Code(Expression(24, Add)) at (prev + 0, 61) to (start + 2, 10)
+ - Code(Expression(21, Add)) at (prev + 0, 61) to (start + 2, 10)
274     = ((c13 + c14) + c15)
275 - Code(Counter(16)) at (prev + 2, 10) to (start + 0, 11)
- - Code(Expression(23, Add)) at (prev + 1, 9) to (start + 1, 18)
```

</details>

Zalathar commented:
I found a way to fix the mappings for `!`: pass an extra `enclosing_not_expr: Option<&Expr<'tcx>>` argument to the recursive call.

We also need to propagate it through the `ExprKind::Scope` case, but the other recursive cases reset it to `None`.

Zalathar commented:
I subsequently realised that my initial “fix” doesn't handle the case of multiple nested nots.

(I was able to get it to preserve the outermost not span, but it was still reversing the then/else arms even when the number of nested nots was even.)

So I switched over to a dedicated struct that separately tracks whether the number of enclosing nots is even or odd.

Zalathar commented:
Another approach might be to stick with `enclosing_not_expr: Option<&Expr>`, but have the coverage code inspect that expression to count how many nots there are between it and the current condition expression.

That would push more of the complexity into coverage-specific code, which might be an improvement overall.

Zalathar committed:
coverage: Record branch information during MIR building

Zalathar commented:
I've switched over to having the coverage-specific code re-derive the number of nested negations by walking down the THIR tree from the outermost enclosing not.

As expected, this is less disruptive for the main lowering code, while still allowing coverage to get the information it needs with little fuss.

bors commented:
:umbrella: The latest upstream changes (presumably #118579) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #118771) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #118500) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Zalathar commented:
Big manual rebase; no longer based on #118237.

bors commented:
:umbrella: The latest upstream changes (presumably #119069) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #119166) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #119122) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Zalathar committed:
coverage: Add branch coverage tests (with branch coverage disabled)

Zalathar commented:
I've made a fairly substantial overhaul [(diff)](https://github.com/rust-lang/rust/compare/3095f69a983a2676cafa8d7ca55aeb10debc12e3..4ddfd882abe8b7381ae96f2ce6953a67dac4ce70).

The most notable change is introducing the `MappingKind` enum, which lets us store regular (“code”) mappings and branch mappings in the same vector, which makes several parts of the code less mistake-prone.

bors commented:
:umbrella: The latest upstream changes (presumably #119648) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #119680) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #119754) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #119767) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #119864) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #120309) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Zalathar committed:
Allow `rustc_mir_transform` to register hook providers

Zalathar committed:
coverage: Make `is_eligible_for_coverage` a hook method

This will allow MIR building to check whether a function is eligible for
coverage instrumentation, and avoid collecting branch coverage info if it is
not.

Zalathar committed:
coverage: Add `CoverageKind::BlockMarker`

Zalathar committed:
coverage: Data structures for recording branch info during MIR building

Zalathar commented:
Big rebase!

I had to resolve some non-trivial merge conflicts, and I also changed the MIR builder code to be specific to branch coverage only, since currently there are no concrete plans to use it for recording non-branch coverage info (though this still remains a future possibility).

Zalathar commented:
The main concern I have with this PR at the moment is that the way it handles `!` expressions is pretty fragile. Because of things like #111752, the current MIR building code makes it a little awkward to determine which arm of an `if` is the true-value arm, and which is the false-value arm.

(The current approach tries to keep track of the outermost enclosing `!`, and then walks back down the THIR tree counting negations until it hits a non-negation node. This avoids major changes to the existing MIR-building code, and gets the job done, but I don't feel good about it.)

My preferred approach would be to rearrange the MIR-building code in a way that naturally satisfies the needs of branch coverage (and beyond), without adversely affecting non-coverage builds. But that's going to require some care, because this code already has some important and subtle responsibilities.

Zalathar commented:
Updated for #121370 (`//@` directives in tests).

Zalathar commented:
I've adjusted how I handle the issue of `||` producing incorrect counts [(diff)](https://github.com/rust-lang/rust/compare/6eb10b81fc5407df7b0916091351a2cd86f430ad..dd09138734b760a4a4944cf50cad9fe33b1f539e).

Instead of sneakily creating an extra block when inserting the block marker, I decided to just fix the issue at its source, by making the `||` lowering create a new block for the LHS and RHS success paths to converge at.

This adds one extra block to the lowering of `||` even when coverage is disabled, but that extra block is soon removed by the MIR simplifier.

Zalathar commented:
I received some questions about this implementation, which I have answered in a Zulip thread: https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/Branch.20coverage.20questions

bors commented:
:umbrella: The latest upstream changes (presumably #121859) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #121998) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #122151) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Zalathar committed:
coverage: Enable branch coverage in the branch coverage tests

Zalathar commented:
Switched to a different mechanism for handling `!` expressions, which is less intrusive to `rustc_mir_build::build::matches`, and should be a bit easier to maintain/extend.

Zalathar commented:
Now that I've resolved my own concerns, I think it might be time to close this and open a new PR for actual review.
